### PR TITLE
State Trie Migration : Concurrent read state trie from old DB

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes

While state trie migration, concurrent read can cut the elapsed time in half.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

### CODE REVIEW

**Comment by KimKyungup** [comment]
I rebased this. PTAL. :)

File: blockchain/blockchain.go [hunk 1]
```diff
@@ -302,32 +320,56 @@ func (bc *BlockChain) migrateState(rootHash common.Hash) error {
...
+	resultCh := make(chan statedb.SyncResult, threads*100)
+
+	for th := 0; th < threads; th++ {
+		go bc.concurrentRead(srcCachedDB, quitCh, hashCh, resultCh)
+	}
+
 	for trieSync.Pending() > 0 {
 		bc.committedCnt, bc.pendingCnt = committedCnt, trieSync.Pending()
-		queue = append(queue[:0], trieSync.Missing(database.IdealBatchSize)...)
+		queue = append(queue[:0], trieSync.Missing(50000)...)
```
Comment (winnie-byun) [inline blockchain/blockchain.go:339]: What do you think about changing this number to `database.IdealBatchSize / threads`?

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -302,32 +320,56 @@ func (bc *BlockChain) migrateState(rootHash common.Hash) error {
 	var queue []common.Hash
 	committedCnt := 0
 
+	quitCh := make(chan struct{})
+	defer close(quitCh)
+
+	// TODO-Klaytn need to tune the threads / batch size
+	//threads := int(math.Max(math.Ceil(float64(runtime.NumCPU())/2), 4))
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:327]: Is there any reason for remaining this comment?

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -302,32 +320,56 @@ func (bc *BlockChain) migrateState(rootHash common.Hash) error {
...
+	threads := int(math.Max(float64(runtime.NumCPU())-2, 4))
+
+	hashCh := make(chan common.Hash, threads*100)
```
Comment (aidan-kwon) [inline blockchain/blockchain.go:330]: I think `threads` is enough instead of `threads*100`.

File: blockchain/blockchain.go (hunk 1 above)
Comment (KimKyungup) [reply blockchain/blockchain.go]: I reverted this code.
Original size shows better performance.

**Pull request merged**